### PR TITLE
Add type annotations and reformat docstring to Google style

### PR DESCRIPTION
**Pull request by DanielYang59**
## Summary

- Add type annotations globally
- Unify docstring to Google style
- Minor clean ups

## Summary by CodeRabbit

- **New Features**
	- Enhanced type hinting across various modules for improved code clarity and maintenance.
	- Introduced clearer documentation and error handling in several areas.
	- Added support for postponed evaluation of type annotations across the library.
- **Documentation**
	- Simplified usage documentation for the `singleton` decorator.
- **Refactor**
	- Refactored variable names and function signatures for better readability.
	- Utilized `contextlib.suppress(Exception)` for cleaner error handling.
- **Style**
	- Implemented `ruff` tool configuration for consistent import sorting.
- **Tests**
	- Updated test files with future import statements for annotations.

### CODE REVIEW

File: monty/string.py [hunk 1]
```diff
@@ -17,7 +19,7 @@ def remove_non_ascii(s):
     return "".join(i for i in s if ord(i) < 128)
 
 
-def is_string(s):
+def is_string(s: Any) -> bool:
```
Comment (DanielYang59) [inline monty/string.py:22]: Not sure what is the advance of this over `isinstance(s, str)`?

**Comment by DanielYang59** [comment]
@coderabbitai review

**Comment by DanielYang59** [comment]
Can you please give me a hand on this remaining `mypy` error? And review this (if you have time)? @janosh Thanks a lot!

**Comment by janosh** [comment]
i think `_exclude_dirs` is possibly unbound if `exclude_dirs` is `None`?

**Comment by DanielYang59** [comment]
> i think `_exclude_dirs` is possibly unbound if `exclude_dirs` is `None`?

Here you are. Thanks! Well I guess it's not... There is a check for `exclude_dirs` before `_exclude_dirs` is used:
```python
        if exclude_dirs is not None and _exclude_dirs.match(mangle(dirpath)):
            continue
        if include_dirs and not _include_dirs.match(mangle(dirpath)):
            continue
```

I have never seen this error before.

**Comment by janosh** [comment]
you're right, it's not actually possibly unbound but i don't think `mypy` can follow the logic here. i would try to not create `_exclude_dirs` and only work with `exclude_dirs`

**Comment by DanielYang59** [comment]
> I would try to not create `_exclude_dirs` and only work with `exclude_dirs`

The original code goes like this. But `exclude_dirs` changed type during run time and of course...`mypy` is not happy:
```python
monty/os/path.py:83: error: Incompatible types in assignment (expression has type "WildCard", variable has type "Optional[str]")  [assignment]
monty/os/path.py:86: error: Incompatible types in assignment (expression has type "WildCard", variable has type "Optional[str]")  [assignment]
monty/os/path.py:95: error: "str" has no attribute "match"  [attr-defined]
monty/os/path.py:95: error: Cannot call function of unknown type  [operator]
monty/os/path.py:97: error: "str" has no attribute "match"  [attr-defined]
monty/os/path.py:97: error: Cannot call function of unknown type  [operator]
```

**Comment by DanielYang59** [comment]
It seems adding type annotations for `mangle` resolves that `mypy` error (88f2b5f6f17bd6e5aa90045bee7182ec6098d87a). Should be ready for review now @janosh. Thanks!

**Review by janosh — CHANGES_REQUESTED** [review]
> It seems adding type annotations for mangle resolves that mypy error (https://github.com/materialsvirtuallab/monty/commit/88f2b5f6f17bd6e5aa90045bee7182ec6098d87a). Should be ready for review now @janosh. Thanks!

nice! fyi, i can't merge on `monty`. has to be done by @shyuep.

this all looks great. only thing i'd suggest is to go straight to future annotations. consider [adding this line](https://github.com/materialsproject/pymatgen/blob/75c771c11605798eb8b3eb026aef865dea3733ec/pyproject.toml#L82) to `pyproject.toml` and then run `ruff`. it'll auto-update all type hints

**Comment by DanielYang59** [comment]
> > It seems adding type annotations for mangle resolves that mypy error ([88f2b5f](https://github.com/materialsvirtuallab/monty/commit/88f2b5f6f17bd6e5aa90045bee7182ec6098d87a)). Should be ready for review now @janosh. Thanks!
> 
> nice! fyi, i can't merge on `monty`. has to be done by @shyuep.

Ahh sorry. Thought you have write access to `monty` too.

> only thing i'd suggest is to go straight to future annotations. consider [adding this line](https://github.com/materialsproject/pymatgen/blob/75c771c11605798eb8b3eb026aef865dea3733ec/pyproject.toml#L82) to `pyproject.toml` and then run `ruff`. it'll auto-update all type hints

Thanks a lot for the useful trick. Tried to add it in 270b3ea48fb37065566dd8e3645b49b36de780a1, but not seeing any effect for some reason?

**Comment by DanielYang59** [comment]
Can you please review this? @shyuep Thanks!

**Comment by janosh** [comment]
you need to select (i.e. enable) the `ruff` `isort` rules too. see the rest of the `pymatgen` ruff config. also `v0.2.1` is old. best update `ruff` with `pre-commit autoupdate`.

**Comment by DanielYang59** [comment]
> you need to select (i.e. enable) the `ruff` `isort` rules too. 

Thanks! I just selected `isort` in 40b5e7b79c3b83c37774fc8d930baca7d35f1285.

> also `v0.2.1` is old. best update `ruff` with `pre-commit autoupdate`.

Done!

**Comment by DanielYang59** [comment]
Can you please review this? @shyuep

**Pull request merged**